specs/derivation.md

```diff
@@ -0,0 +1,537 @@
+
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+TODO
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+# L2 block derivation
+
+[Block derivation][g-derivation] is the process of building a chain of L2 [blocks][g-block] from a chain of L1 data.
+
+The L1 data can be thought of as a stream, and L2 rollup nodes are the ***readers***, transforming their state with the inputs.
+
+The L2 batcher nodes are the ***writers*** to the stream: they take L2 changes and submit the necessary data to L1.
+Others can then reproduce the same L2 changes.
+
+Consuming the stream presents an inductive process:
+given that we know the last L2 block derived from the previous [sequencing window][g-sequencing-window], as well as the
+next [sequencing window][g-sequencing-window], then we can derive [payload attributes] of the next L2 blocks.
+
+To derive the whole L2 chain from scratch, we simply start with the L2 genesis block as the last L2 block, and the
+block at height `L2_CHAIN_INCEPTION + 1` as the start of the next sequencing window.
+Then we iteratively apply the derivation process, using the [derivation pipeline](#derivation-pipeline) process outlined below.
+
+> **TODO** specify genesis block
+
+## Rollup Stream format
+
+The stream is ***multiplexed*** and ***buffered***, to handle serve the required properties:
+
+- There may be different concurrent data submissions to L1
+- There may be different actors that submit the data, the system cannot rely on a single EOA nonce value.
+- The fraud proof processes submitted data, which has to be bounded
+- The submissions confirmed on L1 may be confirmed out of order
+- The L1 may [reorganize][g-reorg], requiring the submissions to reset to a previous point
+- The L2 grows, and does so in smaller intervals than L1
+- The L2 may have bursts that cannot be fully confirmed in a single piece of submitted data
+
+The multiplexed and buffered stream consists of:
+
+- L1 data transactions: any form of L1 data from which frames can be read, tagged with the L1 blockhash, number and timestamp.
+- Channels: independent of each other, a bounded stream submitted by a single actor, encoding a range of L2 bock derivation inputs.
+- Frames: numbered slices of channel data, which can be confirmed on L1 out of order and mixed with other channels.
+
+The types of data are illustrated in the diagram below.
+
+![](./assets/batch-deriv-chain.svg)
+
+### Frame format:
+
+```text
+frame = channel_id | frame_number | frame_data_length | frame_data | is_last
+
+channel_id        = bytes32 | uvarint   # identifies the channel to add frame_data to, as well as the time (UNIX-epoch in seconds) the channel was created.
+frame_number      = uvarint  # identifies the index of frame_data in the channel
+frame_data_length = uvarint  # frame_data_length == len(frame_data)
```

Comment:
I think we should change these to uint32s (could have frame # be uint16, need uint32 for data_length).